File: templates/python/structured_output.py
"""Structured output with Claude Agent SDK (JSON Schema + Pydantic)."""
import asyncio
from pydantic import BaseModel, Field
from claude_agent_sdk import query, ClaudeAgentOptions, ResultMessage


class Issue(BaseModel):
    severity: str = Field(pattern="^(critical|warning|info)$")
    file: str
    line: int | None = None
    description: str


class CodeReview(BaseModel):
    summary: str
    issues: list[Issue]
    score: int = Field(ge=0, le=100)
    recommendation: str = Field(pattern="^(approve|request_changes|needs_discussion)$")


async def main():
    options = ClaudeAgentOptions(
        allowed_tools=["Read", "Grep", "Glob"],
        permission_mode="bypassPermissions",
        output_format={
            "type": "json_schema",
            "schema": CodeReview.model_json_schema(),
        },
    )

    async for msg in query(
        prompt="Review the codebase for security issues and code quality",
        options=options,
    ):
        if isinstance(msg, ResultMessage):
            if msg.subtype == "success" and msg.structured_output:
                review = CodeReview.model_validate(msg.structured_output)
                print(f"Score: {review.score}/100")
                print(f"Recommendation: {review.recommendation}")
                print(f"Issues found: {len(review.issues)}")
                for issue in review.issues:
                    print(f"  [{issue.severity}] {issue.file}:{issue.line or '?'} — {issue.description}")
            elif msg.subtype == "error_max_structured_output_retries":
                print("Agent failed to produce valid structured output after retries")
            elif msg.subtype != "success":
                print(f"Error: {msg.subtype}")


asyncio.run(main())

File: templates/typescript/structured-output.ts
import { query } from "@anthropic-ai/claude-agent-sdk";
import { z } from "zod";
import { toJSONSchema } from "zod/v4/core";

/**
 * Structured Output Template
 *
 * Demonstrates:
 * - JSON Schema output format
 * - Zod schema with draft-07 target
 * - Parsing structured_output from result
 * - Error handling for validation failures
 */

const ReviewSchema = z.object({
  summary: z.string(),
  issues: z.array(
    z.object({
      severity: z.enum(["critical", "warning", "info"]),
      file: z.string(),
      line: z.number().optional(),
      description: z.string(),
    })
  ),
  score: z.number().min(0).max(100),
  recommendation: z.enum(["approve", "request_changes", "needs_discussion"]),
});

async function structuredReview() {
  for await (const message of query({
    prompt: "Review the codebase for security issues and code quality",
    options: {
      allowedTools: ["Read", "Grep", "Glob"],
      permissionMode: "bypassPermissions",
      allowDangerouslySkipPermissions: true,
      outputFormat: {
        type: "json_schema",
        // IMPORTANT: Use target: "draft-07" — Claude requires draft-07, not draft-2020-12
        schema: toJSONSchema(ReviewSchema, { target: "draft-07" }),
      },
    },
  })) {
    switch (message.type) {
      case "result":
        if (message.subtype === "success" && message.structured_output) {
          const parsed = ReviewSchema.safeParse(message.structured_output);
          if (parsed.success) {
            console.log(`Score: ${parsed.data.score}/100`);
            console.log(`Recommendation: ${parsed.data.recommendation}`);
            console.log(`Issues found: ${parsed.data.issues.length}`);
            for (const issue of parsed.data.issues) {
              console.log(`  [${issue.severity}] ${issue.file}:${issue.line ?? "?"} — ${issue.description}`);
            }
          } else {
            console.error("Schema validation failed:", parsed.error);
          }
        } else if (message.subtype === "error_max_structured_output_retries") {
          console.error("Agent failed to produce valid structured output after retries");
        } else if (message.subtype !== "success") {
          console.error(`Error: ${message.subtype}`, message.errors);
        }
        break;
    }
  }
}

structuredReview().catch(console.error);
